File: js/packages/sdk/mcp.ts
/**
 * Durable MCP (Model Context Protocol) middleware for Restate.
 * Provides a middleware that makes all MCP fetch operations replayable via Restate.
 */

import { Client, type ClientOptions } from "@modelcontextprotocol/sdk/client";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import type { RequestOptions } from "@modelcontextprotocol/sdk/shared/protocol.js";
import type { Transport } from "@modelcontextprotocol/sdk/shared/transport.js";
import type {
	CallToolRequest,
	CallToolResultSchema,
	CompatibilityCallToolResultSchema,
	GetPromptRequest,
	Implementation,
	ListPromptsRequest,
	ListResourcesRequest,
	ListResourceTemplatesRequest,
	ListToolsRequest,
	Notification,
	ReadResourceRequest,
	Result,
	SubscribeRequest,
	UnsubscribeRequest,
} from "@modelcontextprotocol/sdk/types.js";
import type { Context, ObjectContext } from "@restatedev/restate-sdk";

/**
 * Configuration options for creating a Soma MCP client.
 */
export interface SomaMcpClientConfig {
	/**
	 * The base URL of the Soma server.
	 * Defaults to SOMA_SERVER_BASE_URL env var or 'http://localhost:3000'.
	 */
	somaBaseUrl?: string;
}

/**
 * Creates an MCP client connected to a Soma MCP server instance.
 *
 * @param mcpServerInstanceId - The ID of the MCP server instance to connect to.
 * @param config - Optional configuration including base URL.
 * @returns A connected MCP Client instance.
 *
 * @example
 * ```typescript
 * import { createSomaMcpClient } from '@trysoma/sdk/mcp';
 *
 * const client = await createSomaMcpClient('my-mcp-instance-id');
 *
 * // List available tools
 * const tools = await client.listTools();
 *
 * // Call a tool
 * const result = await client.callTool({ name: 'my-tool', arguments: {} });
 * ```
 */
export async function createSomaMcpClient<RequestT extends Request = Request, NotificationT extends Notification = Notification, ResultT extends Result = Result>(
	ctx: ObjectContext,
	mcpServerInstanceId: string,
	config?: SomaMcpClientConfig,
): Promise<SomaMcpClient<RequestT, NotificationT, ResultT>> {


	const client = new SomaMcpClient<RequestT, NotificationT, ResultT>(
		ctx,
		{
			name: mcpServerInstanceId,
			version: "1.0.0",
		},
	);

	const { transport, sessionId: _sessionId } = createSomaMcpTransport(
		ctx,
		mcpServerInstanceId,
		config,
	);

	// Set up error handler before connecting (matches frontend pattern)
	client.onerror = (error) => {
		ctx.console.error("MCP client error:", error);
	};

	// Set up close handler before connecting (matches frontend pattern)
	client.onclose = () => {
		ctx.console.debug("MCP client connection closed");
	};

	// Create fresh connection - don't wrap in ctx.run to avoid session resumption issues
	// The connection is ephemeral and will be recreated on each invocation/replay
	// If connection fails due to session resumption, the error handler will catch it
	try {
		await client.connect(transport);
	} catch (error) {
		// If connection fails (e.g., due to session resumption issues), log and rethrow
		// The error handler above will also catch it
		ctx.console.error("Failed to connect MCP client:", error);
		throw error;
	}
	return client;
}

export function createSomaMcpTransport(
	_ctx: Context,
	mcpServerInstanceId: string,
	config?: SomaMcpClientConfig,
): { transport: StreamableHTTPClientTransport; sessionId: string | undefined } {
	const baseUrl =
		config?.somaBaseUrl ||
		// biome-ignore lint/complexity/useLiteralKeys: TypeScript requires bracket notation for process.env
		process.env["SOMA_SERVER_BASE_URL"] ||
		"http://localhost:3000";

	const mcpUrl = new URL(
		`/api/bridge/v1/mcp-instance/${mcpServerInstanceId}/mcp`,
		baseUrl,
	);

	let sessionId: string | undefined;
	const transport = new StreamableHTTPClientTransport(mcpUrl);

	return {
		transport,
		sessionId,
	};
}

export class SomaMcpClient<
	RequestT extends Request = Request,
	NotificationT extends Notification = Notification,
	ResultT extends Result = Result,
> extends Client<RequestT, NotificationT, ResultT> {
	private restate: ObjectContext;
	private clientInfo: Implementation;

	public constructor(
		restate: ObjectContext,
		clientInfo: Implementation,
		options?: ClientOptions,
	) {
		super(clientInfo, options);
		this.restate = restate;
		this.clientInfo = clientInfo;
	}

	override async connect(
		transport: Transport,
		options?: RequestOptions,
	): Promise<void> {
		return this.restate.run(`mcp-${this.clientInfo.name}-connect`, async () => {
			return super.connect(transport, options);
		});
	}

	override async ping(options?: RequestOptions): Promise<{
		_meta?: Record<string, unknown> | undefined;
	}> {
		return this.restate.run(`mcp-${this.clientInfo.name}-ping`, async () => {
			return super.ping(options);
		});
	}

	override async getPrompt(
		params: GetPromptRequest["params"],
		options?: RequestOptions,
	): Promise<{
		[x: string]: unknown;
		messages: {
			role: "user" | "assistant";
			content:
				| {
						type: "text";
						text: string;
						_meta?: Record<string, unknown> | undefined;
				  }
				| {
						type: "image";
						data: string;
						mimeType: string;
						_meta?: Record<string, unknown> | undefined;
				  }
				| {
						type: "audio";
						data: string;
						mimeType: string;
						_meta?: Record<string, unknown> | undefined;
				  }
				| {
						type: "resource";
						resource:
							| {
									uri: string;
									text: string;
									mimeType?: string | undefined;
									_meta?: Record<string, unknown> | undefined;
							  }
							| {
									uri: string;
									blob: string;
									mimeType?: string | undefined;
									_meta?: Record<string, unknown> | undefined;
							  };
						_meta?: Record<string, unknown> | undefined;
				  }
				| {
						uri: string;
						name: string;
						type: "resource_link";
						description?: string | undefined;
						mimeType?: string | undefined;
						_meta?:
							| {
									[x: string]: unknown;
							  }
							| undefined;
						icons?:
							| {
									src: string;
									mimeType?: string | undefined;
									sizes?: string[] | undefined;
							  }[]
							| undefined;
						title?: string | undefined;
				  };
		}[];
		_meta?: Record<string, unknown> | undefined;
		description?: string | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-getPrompt`,
			async () => {
				return super.getPrompt(params, options);
			},
		);
	}
	override async listPrompts(
		params?: ListPromptsRequest["params"],
		options?: RequestOptions,
	): Promise<{
		[x: string]: unknown;
		prompts: {
			name: string;
			description?: string | undefined;
			arguments?:
				| {
						name: string;
						description?: string | undefined;
						required?: boolean | undefined;
				  }[]
				| undefined;
			_meta?:
				| {
						[x: string]: unknown;
				  }
				| undefined;
			icons?:
				| {
						src: string;
						mimeType?: string | undefined;
						sizes?: string[] | undefined;
				  }[]
				| undefined;
			title?: string | undefined;
		}[];
		_meta?: Record<string, unknown> | undefined;
		nextCursor?: string | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-listPrompts`,
			async () => {
				return super.listPrompts(params, options);
			},
		);
	}
	override async listResources(
		params?: ListResourcesRequest["params"],
		options?: RequestOptions,
	): Promise<{
		[x: string]: unknown;
		resources: {
			uri: string;
			name: string;
			description?: string | undefined;
			mimeType?: string | undefined;
			_meta?:
				| {
						[x: string]: unknown;
				  }
				| undefined;
			icons?:
				| {
						src: string;
						mimeType?: string | undefined;
						sizes?: string[] | undefined;
				  }[]
				| undefined;
			title?: string | undefined;
		}[];
		_meta?: Record<string, unknown> | undefined;
		nextCursor?: string | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-listResources`,
			async () => {
				return super.listResources(params, options);
			},
		);
	}
	override async listResourceTemplates(
		params?: ListResourceTemplatesRequest["params"],
		options?: RequestOptions,
	): Promise<{
		[x: string]: unknown;
		resourceTemplates: {
			uriTemplate: string;
			name: string;
			description?: string | undefined;
			mimeType?: string | undefined;
			_meta?:
				| {
						[x: string]: unknown;
				  }
				| undefined;
			icons?:
				| {
						src: string;
						mimeType?: string | undefined;
						sizes?: string[] | undefined;
				  }[]
				| undefined;
			title?: string | undefined;
		}[];
		_meta?: Record<string, unknown> | undefined;
		nextCursor?: string | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-listResourceTemplates`,
			async () => {
				return super.listResourceTemplates(params, options);
			},
		);
	}
	override async readResource(
		params: ReadResourceRequest["params"],
		options?: RequestOptions,
	): Promise<{
		[x: string]: unknown;
		contents: (
			| {
					uri: string;
					text: string;
					mimeType?: string | undefined;
					_meta?: Record<string, unknown> | undefined;
			  }
			| {
					uri: string;
					blob: string;
					mimeType?: string | undefined;
					_meta?: Record<string, unknown> | undefined;
			  }
		)[];
		_meta?: Record<string, unknown> | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-readResource`,
			async () => {
				return super.readResource(params, options);
			},
		);
	}
	override async subscribeResource(
		params: SubscribeRequest["params"],
		options?: RequestOptions,
	): Promise<{
		_meta?: Record<string, unknown> | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-subscribeResource`,
			async () => {
				return super.subscribeResource(params, options);
			},
		);
	}
	override async unsubscribeResource(
		params: UnsubscribeRequest["params"],
		options?: RequestOptions,
	): Promise<{
		_meta?: Record<string, unknown> | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-unsubscribeResource`,
			async () => {
				return super.unsubscribeResource(params, options);
			},
		);
	}
	override async callTool(
		params: CallToolRequest["params"],
		resultSchema?:
			| typeof CallToolResultSchema
			| typeof CompatibilityCallToolResultSchema,
		options?: RequestOptions,
	): Promise<
		| {
				[x: string]: unknown;
				content: (
					| {
							type: "text";
							text: string;
							_meta?: Record<string, unknown> | undefined;
					  }
					| {
							type: "image";
							data: string;
							mimeType: string;
							_meta?: Record<string, unknown> | undefined;
					  }
					| {
							type: "audio";
							data: string;
							mimeType: string;
							_meta?: Record<string, unknown> | undefined;
					  }
					| {
							type: "resource";
							resource:
								| {
										uri: string;
										text: string;
										mimeType?: string | undefined;
										_meta?: Record<string, unknown> | undefined;
								  }
								| {
										uri: string;
										blob: string;
										mimeType?: string | undefined;
										_meta?: Record<string, unknown> | undefined;
								  };
							_meta?: Record<string, unknown> | undefined;
					  }
					| {
							uri: string;
							name: string;
							type: "resource_link";
							description?: string | undefined;
							mimeType?: string | undefined;
							_meta?:
								| {
										[x: string]: unknown;
								  }
								| undefined;
							icons?:
								| {
										src: string;
										mimeType?: string | undefined;
										sizes?: string[] | undefined;
								  }[]
								| undefined;
							title?: string | undefined;
					  }
				)[];
				_meta?: Record<string, unknown> | undefined;
				structuredContent?: Record<string, unknown> | undefined;
				isError?: boolean | undefined;
		  }
		| {
				[x: string]: unknown;
				toolResult: unknown;
				_meta?: Record<string, unknown> | undefined;
		  }
	> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-callTool`,
			async () => {
				return super.callTool(params, resultSchema, options);
			},
		);
	}

	override async listTools(
		params?: ListToolsRequest["params"],
		options?: RequestOptions,
	): Promise<{
		[x: string]: unknown;
		tools: {
			inputSchema: {
				[x: string]: unknown;
				type: "object";
				properties?: Record<string, object> | undefined;
				required?: string[] | undefined;
			};
			name: string;
			description?: string | undefined;
			outputSchema?:
				| {
						[x: string]: unknown;
						type: "object";
						properties?: Record<string, object> | undefined;
						required?: string[] | undefined;
				  }
				| undefined;
			annotations?:
				| {
						title?: string | undefined;
						readOnlyHint?: boolean | undefined;
						destructiveHint?: boolean | undefined;
						idempotentHint?: boolean | undefined;
						openWorldHint?: boolean | undefined;
				  }
				| undefined;
			_meta?: Record<string, unknown> | undefined;
			icons?:
				| {
						src: string;
						mimeType?: string | undefined;
						sizes?: string[] | undefined;
				  }[]
				| undefined;
			title?: string | undefined;
		}[];
		_meta?: Record<string, unknown> | undefined;
		nextCursor?: string | undefined;
	}> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-listTools`,
			async () => {
				return super.listTools(params, options);
			},
		);
	}
	override async sendRootsListChanged(): Promise<void> {
		return this.restate.run(
			`mcp-${this.clientInfo.name}-sendRootsListChanged`,
			async () => {
				return super.sendRootsListChanged();
			},
		);
	}
}

File: py/packages/sdk/trysoma_sdk/mcp.py
"""
Durable MCP (Model Context Protocol) client for Restate.
Provides a client wrapper that makes all MCP operations replayable via Restate.
"""

import logging
import os
from contextlib import asynccontextmanager
from collections.abc import AsyncIterator
from typing import cast

logger = logging.getLogger(__name__)

from mcp import ClientSession
from mcp.client.streamable_http import streamablehttp_client
from mcp.types import (
    CallToolResult,
    EmptyResult,
    GetPromptResult,
    ListPromptsResult,
    ListResourcesResult,
    ListResourceTemplatesResult,
    ListToolsResult,
    ReadResourceResult,
)
from pydantic import AnyUrl
from restate import ObjectContext


class SomaMcpClientConfig:
    """Configuration options for creating a Soma MCP client."""

    def __init__(self, soma_base_url: str | None = None) -> None:
        """
        Initialize config.

        Args:
            soma_base_url: The base URL of the Soma server.
                          Defaults to SOMA_SERVER_BASE_URL env var or 'http://localhost:3000'.
        """
        self.soma_base_url = soma_base_url


def get_mcp_url(
    mcp_server_instance_id: str,
    config: SomaMcpClientConfig | None = None,
) -> str:
    """
    Get the MCP URL for a Soma MCP server instance.

    Args:
        mcp_server_instance_id: The ID of the MCP server instance to connect to.
        config: Optional configuration including base URL.

    Returns:
        The MCP server URL.
    """
    base_url = (
        config.soma_base_url
        if config and config.soma_base_url
        else os.environ.get("SOMA_SERVER_BASE_URL", "http://localhost:3000")
    )

    return f"{base_url}/api/bridge/v1/mcp-instance/{mcp_server_instance_id}/mcp"


class SomaMcpClient:
    """
    Durable MCP client wrapper that makes all MCP operations replayable via Restate.

    The client maintains a persistent connection to the MCP server and wraps
    all operations with ctx.run() for durability.
    """

    def __init__(
        self,
        ctx: ObjectContext,
        session: ClientSession,
        client_name: str,
    ) -> None:
        """
        Initialize the durable MCP client.

        Args:
            ctx: The Restate ObjectContext for durability
            session: The MCP ClientSession (must be initialized)
            client_name: A unique name for this client (used for durability keys)
        """
        self._ctx = ctx
        self._session = session
        self._client_name = client_name
        self._request_index = 0

    async def ping(self) -> EmptyResult:
        """Ping the MCP server (durable)."""
        logger.debug("Pinging MCP server %s", self._client_name)
        index = self._request_index
        self._request_index += 1

        async def do_ping() -> dict[str, object]:
            result = await self._session.send_ping()
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-ping-index-{index}", do_ping
        )
        logger.debug("Pinging MCP server %s completed", self._client_name)
        return EmptyResult.model_validate(data)

    async def list_tools(
        self,
        cursor: str | None = None,
    ) -> ListToolsResult:
        """List available tools (durable)."""
        logger.debug("Listing tools on MCP server %s", self._client_name)
        index = self._request_index
        self._request_index += 1

        async def do_list() -> dict[str, object]:
            result = await self._session.list_tools(cursor=cursor)
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-listTools-index-{index}", do_list
        )
        result = ListToolsResult.model_validate(data)
        logger.debug("Listing tools on MCP server %s completed (count=%d)", self._client_name, len(result.tools))
        return result

    async def call_tool(
        self,
        name: str,
        arguments: dict[str, object] | None = None,
    ) -> CallToolResult:
        """Call a tool on the MCP server (durable)."""
        logger.debug("Calling tool %s on MCP server %s", name, self._client_name)
        index = self._request_index
        self._request_index += 1

        async def do_call() -> dict[str, object]:
            result = await self._session.call_tool(name, arguments or {})
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-callTool-index-{index}", do_call
        )
        result = CallToolResult.model_validate(data)
        logger.debug("Calling tool %s on MCP server %s completed", name, self._client_name)
        return result

    async def list_resources(
        self,
        cursor: str | None = None,
    ) -> ListResourcesResult:
        """List available resources (durable)."""
        logger.debug("Listing resources on MCP server %s", self._client_name)
        index = self._request_index
        self._request_index += 1

        async def do_list() -> dict[str, object]:
            result = await self._session.list_resources(cursor=cursor)
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-listResources-index-{index}", do_list
        )
        result = ListResourcesResult.model_validate(data)
        logger.debug("Listing resources on MCP server %s completed", self._client_name)
        return result

    async def list_resource_templates(
        self,
        cursor: str | None = None,
    ) -> ListResourceTemplatesResult:
        """List available resource templates (durable)."""
        index = self._request_index
        self._request_index += 1

        async def do_list() -> dict[str, object]:
            result = await self._session.list_resource_templates(cursor=cursor)
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-listResourceTemplates-index-{index}", do_list
        )
        return ListResourceTemplatesResult.model_validate(data)

    async def read_resource(self, uri: str) -> ReadResourceResult:
        """Read a resource by URI (durable)."""
        logger.debug("Reading resource %s on MCP server %s", uri, self._client_name)
        index = self._request_index
        self._request_index += 1

        async def do_read() -> dict[str, object]:
            result = await self._session.read_resource(AnyUrl(uri))
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-readResource-index-{index}", do_read
        )
        result = ReadResourceResult.model_validate(data)
        logger.debug("Reading resource %s on MCP server %s completed", uri, self._client_name)
        return result

    async def subscribe_resource(self, uri: str) -> EmptyResult:
        """Subscribe to a resource for updates (durable)."""
        index = self._request_index
        self._request_index += 1

        async def do_subscribe() -> dict[str, object]:
            result = await self._session.subscribe_resource(AnyUrl(uri))
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-subscribeResource-index-{index}", do_subscribe
        )
        return EmptyResult.model_validate(data)

    async def unsubscribe_resource(self, uri: str) -> EmptyResult:
        """Unsubscribe from a resource (durable)."""
        index = self._request_index
        self._request_index += 1

        async def do_unsubscribe() -> dict[str, object]:
            result = await self._session.unsubscribe_resource(AnyUrl(uri))
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-unsubscribeResource-index-{index}", do_unsubscribe
        )
        return EmptyResult.model_validate(data)

    async def list_prompts(
        self,
        cursor: str | None = None,
    ) -> ListPromptsResult:
        """List available prompts (durable)."""
        index = self._request_index
        self._request_index += 1

        async def do_list() -> dict[str, object]:
            result = await self._session.list_prompts(cursor=cursor)
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-listPrompts-index-{index}", do_list
        )
        return ListPromptsResult.model_validate(data)

    async def get_prompt(
        self,
        name: str,
        arguments: dict[str, str] | None = None,
    ) -> GetPromptResult:
        """Get a prompt by name (durable)."""
        logger.debug("Getting prompt %s on MCP server %s", name, self._client_name)
        index = self._request_index
        self._request_index += 1

        async def do_get() -> dict[str, object]:
            result = await self._session.get_prompt(name, arguments)
            return cast(dict[str, object], result.model_dump())

        data = await self._ctx.run(
            f"mcp-{self._client_name}-getPrompt-index-{index}", do_get
        )
        result = GetPromptResult.model_validate(data)
        logger.debug("Getting prompt %s on MCP server %s completed", name, self._client_name)
        return result

    async def send_roots_list_changed(self) -> None:
        """Send roots list changed notification (durable)."""
        index = self._request_index
        self._request_index += 1

        async def do_send() -> None:
            await self._session.send_roots_list_changed()

        await self._ctx.run(
            f"mcp-{self._client_name}-sendRootsListChanged-index-{index}", do_send
        )


@asynccontextmanager
async def create_soma_mcp_client(
    ctx: ObjectContext,
    mcp_server_instance_id: str,
    config: SomaMcpClientConfig | None = None,
) -> AsyncIterator[SomaMcpClient]:
    """
    Create an MCP client connected to a Soma MCP server instance.

    This is an async context manager that maintains the connection for the
    duration of the context. Use it with `async with`:

    Args:
        ctx: The Restate ObjectContext for durability
        mcp_server_instance_id: The ID of the MCP server instance to connect to.
        config: Optional configuration including base URL.

    Yields:
        A connected SomaMcpClient instance.

    Example:
        ```python
        from trysoma_sdk.mcp import create_soma_mcp_client

        async with create_soma_mcp_client(ctx, 'my-mcp-instance-id') as client:
            # List available tools
            tools = await client.list_tools()

            # Call a tool
            result = await client.call_tool('my-tool', {'arg': 'value'})
        ```
    """
    logger.debug("Creating MCP client for instance %s", mcp_server_instance_id)
    mcp_url = get_mcp_url(mcp_server_instance_id, config)

    async with streamablehttp_client(mcp_url) as (
        read_stream,
        write_stream,
        _get_session_id,
    ):
        async with ClientSession(
            read_stream=read_stream,
            write_stream=write_stream,
        ) as session:
            await session.initialize()
            logger.debug("Creating MCP client for instance %s completed", mcp_server_instance_id)
            yield SomaMcpClient(ctx, session, mcp_server_instance_id)


# Re-export for convenience
__all__ = [
    # Main client
    "SomaMcpClient",
    "SomaMcpClientConfig",
    "create_soma_mcp_client",
    "get_mcp_url",
    # Re-exported types
    "CallToolResult",
    "ListToolsResult",
    "ListResourcesResult",
    "ListResourceTemplatesResult",
    "ReadResourceResult",
    "ListPromptsResult",
    "GetPromptResult",
    "EmptyResult",
]
